parlai/core/testing_utils.py

```diff
@@ -0,0 +1,231 @@
+#!/usr/bin/env python
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+General utilities for helping writing ParlAI unit and integration tests.
+"""
+
+import os
+import unittest
+import contextlib
+import tempfile
+import shutil
+import io
+
+
+try:
+    import torch
+    TORCH_AVAILABLE = True
+    GPU_AVAILABLE = torch.cuda.device_count() > 0
+except ImportError:
+    TORCH_AVAILABLE = False
+    GPU_AVAILABLE = False
+
+try:
+    import git
+    git_ = git.Git()
+    GIT_AVAILABLE = True
+except ImportError:
+    git_ = None
+    GIT_AVAILABLE = False
+
+
+DEBUG = False  # change this to true to print to stdout anyway
+
+
+def skipUnlessTorch(testfn, reason='pytorch is not installed'):
+    """Decorator for skipping a test if torch is not installed."""
+    return unittest.skipUnless(TORCH_AVAILABLE, reason)(testfn)
+
+
+def skipIfGPU(testfn, reason='Test is CPU-only'):
+    """
+    Decorator for skipping a test if a GPU is available.
+
+    Useful for disabling hogwild tests.
+    """
+    return unittest.skipIf(GPU_AVAILABLE, reason)(testfn)
+
+
+def skipUnlessGPU(testfn, reason='Test requires a GPU'):
+    """Decorator for skipping a test if no GPU is available."""
+    return unittest.skipIf(not GPU_AVAILABLE, reason)(testfn)
```

Comment:
nit: considering you used the `skipUnless` syntax above, maybe good to be consistent?